src/tools/rbd_mirror/LeaderWatcher.h

```diff
@@ -4,32 +4,76 @@
 #ifndef CEPH_RBD_MIRROR_LEADER_WATCHER_H
 #define CEPH_RBD_MIRROR_LEADER_WATCHER_H
 
+#include "librbd/ManagedLock.h"
+#include "librbd/managed_lock/Types.h"
 #include "librbd/Watcher.h"
+#include "MirrorStatusWatcher.h"
 #include "tools/rbd_mirror/leader_watcher/Types.h"
 
+namespace librbd {
+  class ImageCtx;
+}
+
 namespace rbd {
 namespace mirror {
 
 class LeaderWatcher : protected librbd::Watcher {
 public:
-  LeaderWatcher(librados::IoCtx &io_ctx, ContextWQ *work_queue);
+  LeaderWatcher(librados::IoCtx &io_ctx, ContextWQ *work_queue, Mutex &lock,
+                Cond &cond);
 
   int init();
   void shut_down();
 
+  bool is_leader() {
+    assert(m_lock.is_locked());
+    return !m_stopping && m_leader_lock && m_leader_lock->is_lock_owner();
+  }
+
+  void check_leader_alive(utime_t &now, int heartbeat_interval);
+
   void notify_heartbeat(Context *on_finish);
   void notify_lock_acquired(Context *on_finish);
```

Comment:
Do these two methods (notify_lock_acquired  and notify_lock_released ) need to be public? You have overloaded versions that interact w/ the internal state machine and these stand-alone ones.